### PR TITLE
Refactor librustdoc html backend

### PR DESCRIPTION
**Pull request by P1n3appl3**
This PR moves several types out of the librustdoc::html module so that they can be used by a future json backend. These changes are a re-implementation of [some work done 6 months ago](https://github.com/rust-lang/rust/compare/master...GuillaumeGomez:multiple-output-formats) by @GuillaumeGomez. I'm currently working on said json backend and will put up an RFC soon with the proposed implementation.

There are a couple of changes that are more substantial than relocating structs to a different module:
1. The `Cache` is no longer part of the `html::render::Context` type and therefor it needs to be explicitly passed to any functions that access it.
2. The driving function `html::render::run` has been rewritten to use the `FormatRenderer` trait which should allow different backends to re-use the driving code.

r? @GuillaumeGomez

cc @tmandry @betamos

### CODE REVIEW

File: src/librustdoc/html/render.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: Since you're passing `diag` down to the other methods, I don't think we need to keep it in here.

File: src/librustdoc/html/render.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: Just a note (for myself and any other potential reviewer): those 3 variables were before in the `run` function which became split in the `FormatRenderer` trait.

File: src/librustdoc/formats/mod.rs
Comment (GuillaumeGomez) [inline src/librustdoc/formats/mod.rs]: Note: this allows to check if any errors has occurred during the run, which allows to make the code inside `after_krate` a bit lighter.

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Please just remove the `diag` parameter where it's unused. Otherwise looks good to me, thanks for working on this!

File: src/librustdoc/formats/mod.rs
Comment (betamos) [inline src/librustdoc/formats/mod.rs]: Maybe just "should not" instead of "doesn't need to"?

File: src/librustdoc/formats/mod.rs
Comment (betamos) [inline src/librustdoc/formats/mod.rs]: Just curious, would this ever be anything different than Self? Does it need to be an associated type?

File: src/librustdoc/formats/mod.rs
Comment (betamos) [inline src/librustdoc/formats/mod.rs]: Could you document public items (here, the trait itself and the init function)? That said, I'm fresh in this code base so maybe this convention doesn't apply.

File: src/librustdoc/lib.rs
Comment (betamos) [inline src/librustdoc/lib.rs:69]: Is librustdoc meant to be dependended on from other repos? Changing visibility of a previously exported module is quite drastic so just want to make.

File: src/librustdoc/formats/mod.rs
Comment (betamos) [inline src/librustdoc/formats/mod.rs:12]: Here and below: a short docstring would be helpful.

**Review by tmandry — COMMENTED** [review]
Looks good! I left a few nits, feel free to ignore if they are preexisting.

File: src/librustdoc/lib.rs
Comment (jyn514) [reply src/librustdoc/lib.rs:69]: AFAIK rustdoc is only used as a library from `src/tools/rustdoc`. I would be very surprised if anyone else was depending on it.

**Review by jyn514 — COMMENTED** [review]
I like the idea but this is going to conflict with every one of the intra-doc link PRs 😅

File: src/librustdoc/lib.rs
Comment (bjorn3) [reply src/librustdoc/lib.rs:69]: librustdoc is not part of the sysroot, so you will need to compile it yourself if you want to use it.

File: src/librustdoc/formats/mod.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/mod.rs]: Nope, good catch. That was an artifact of a previous version.

**Comment by P1n3appl3** [comment]
@jyn514 I couldn't find those PRs, do you mean they'll conflict because of moving `html/render.rs` to `html/render/mod.rs` or because of the changes in general? I could just keep the file in the same place if it helps.

File: src/librustdoc/lib.rs
Comment (P1n3appl3) [reply src/librustdoc/lib.rs]: it turns out making it `crate` breaks a doc compiletest in `html/markdown.rs` so i had to revert it to `pub` anyway

**Comment by jyn514** [comment]
Err I was actually thinking of https://github.com/rust-lang/rust/pull/72771, https://github.com/rust-lang/rust/pull/73101 doesn't actually modify `html/` at all. So no need to change anything. Thanks for the offer though!

**Review by GuillaumeGomez — APPROVED** [review]
Looks good to me! Considering this is a huge PR (even if mainly doing code moves), I'll wait for at least one more review from someone on the @rust-lang/rustdoc team (ping @ollie27 )

**Comment by tmandry** [comment]
@ollie27 Friendly ping, would you be able to review this?

**Comment by tmandry** [comment]
Perhaps another person on @rust-lang/rustdoc can give this a once-over. I nominate...

r? @Manishearth

**Comment by Manishearth** [comment]
@jyn514 wanna review this?

File: src/librustdoc/formats/mod.rs
Comment (Manishearth) [inline src/librustdoc/formats/mod.rs]: Feels like this should just be a toplevel `fn run_format` function, singleton classes aren't as much of a thing in rust

**Review by Manishearth — APPROVED** [review]
minor nit, otherwise r=me!

File: src/librustdoc/formats/cache.rs
Comment (jyn514) [inline src/librustdoc/formats/cache.rs:214]:
```suggestion
                    cache.impls.entry(did).or_default().push(impl_.clone());
```

File: src/librustdoc/formats/renderer.rs
Comment (jyn514) [inline src/librustdoc/formats/renderer.rs:104]: Cloning this for each _item_ seems like a waste :/

File: src/librustdoc/formats/renderer.rs
Comment (jyn514) [inline src/librustdoc/formats/renderer.rs:54]: This seems redundant since `FormatRenderer: Clone`.
```suggestion
    pub fn run<T: FormatRenderer>(
```

File: src/librustdoc/formats/renderer.rs
Comment (jyn514) [inline src/librustdoc/formats/renderer.rs:46]: This struct seems unnecessary. Could this be `formats::renderer::run` instead?

File: src/librustdoc/html/render/mod.rs
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:202]: If you make this fully `pub`, rustc won't warn about dead code.
```suggestion
pub(crate) struct IndexItem {
```

File: src/librustdoc/formats/cache.rs
Comment (jyn514) [inline src/librustdoc/formats/cache.rs:101]: It looks like there is both a `formats::cache` and a `html::render::cache`. Maybe this belongs with `render`? I don't see why backends other than HTML would need a JS search index.

File: src/librustdoc/html/render/mod.rs
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:231]:
```suggestion
pub(crate) struct RenderType {
```

File: src/librustdoc/html/render/mod.rs
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:262]:
```suggestion
pub(crate) struct Generic {
```

File: src/librustdoc/html/render/mod.rs
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:283]:
```suggestion
pub(crate) struct IndexItemFunctionType {
```

**Review by jyn514 — COMMENTED** [review]
Most of these are about pre-existing issues, feel free to merge without addressing.

File: src/librustdoc/formats/mod.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/mod.rs]: Good point, the idea was for that struct to hold state shared by the other renderers but the cache covers that pretty well so I'll remove it.

Comment (Manishearth) [reply src/librustdoc/formats/mod.rs]: It might be worth eventually making the caching explicit as opposed to being a global mutable, but that can happen elsewhere

File: src/librustdoc/formats/cache.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/cache.rs:40]: Nope, don't see std or other types in there unless i'm reexporting them. It does have the numeric primitives + char and bool but I assume that's either from the prelude or just because they're special.

Comment (jyn514) [reply src/librustdoc/formats/cache.rs:40]: Got it. In general, if it doesn't have types from other crates it should use `LocalDefId` but that can be delayed to a later PR.

**Comment by jyn514** [comment]
I know I left a giant stream of comments - most of those are from me reading the codebase the first time and my first impressions. So please don't block on those.

**Comment by jyn514** [comment]
(except the `pub(crate)` thing which I do think is useful and related to this change)

**Comment by P1n3appl3** [comment]
Yeah i wasn't blocked, should have posted my draft responses.The only other change I'm making is removing the cloning of `Context` for every item. Running tests now (which is taking a bit since I rebased and have to recompile) and then I'll be done.

File: src/librustdoc/formats/cache.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/cache.rs:47]: Probably, although it's kind of convenient that all the mappings in `Cache` use a full `DefId`. Otherwise I think you'd need to add checks that a given item is local before pulling out it's `LocalDefId` and looking it up in those maps.

File: src/librustdoc/formats/renderer.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/renderer.rs:104]: Yeah I thought the same thing. It's how @GuillaumeGomez implemented it last year and I assumed it was to circumvent ownership issues, but looking at it again it should be fine to share a mutable copy (and if we ever want to parallelize item rendering we could put it behind a mutex) so I'll go ahead and remove the cloning.

File: src/librustdoc/formats/cache.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/cache.rs:101]: Yeah it would definitely be better to have `search_index` in `html::render::SharedContext` but it requires a bit of work to separate out the logic to construct the index vs the cache because they're currently using the same `DocFolder` pass.

File: src/librustdoc/html/render/mod.rs
Comment (P1n3appl3) [reply src/librustdoc/html/render/mod.rs:283]: This one gets exposed by the cache so it's got to stay public (until we can move search index related stuff to `html::render::SharedContext`)

File: src/librustdoc/formats/renderer.rs
Comment (P1n3appl3) [reply src/librustdoc/formats/renderer.rs:104]: Shoot nevermind, I just remembered why this was necessary (after seeing hundreds of tests fail): the html renderer maintains a mutable "current" path which updates as it traverses submodules. It makes the output depend on the order in which items get rendered and cloning the `Context` fixes that by preserving a "snapshot" of that current path for each item. When I've got a bit of time I'll refactor the html renderer to not do that and then we can remove this cloning as well, but I'm reverting for now though.

File: src/librustdoc/html/render/mod.rs
Comment (P1n3appl3) [reply src/librustdoc/html/render/mod.rs:202]: Can't restrict it since it's exposed through `Cache`

**Comment by tmandry** [comment]
Looks like all comments have been addressed, so

@bors r+

**Comment by Manishearth** [comment]
A bajillion test failures of the form:

```
2020-07-28T03:35:02.5974142Z thread '<unnamed>' panicked at 'failed to send error on "D:\a\rust\rust\build\x86_64-pc-windows-msvc\test\rustdoc\traits-in-bodies-private\traits_in_bodies_private\struct.SomeStruct.html"', src\librustdoc\docfs.rs:97:25
2020-07-28T03:35:02.5974376Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: src/librustdoc/docfs.rs
Comment (tmandry) [inline src/librustdoc/docfs.rs:39]: nit: better to pass the sender by value, the caller shouldn't need it

**Comment by jyn514** [comment]
@bors rollup=iffy

This will cause lots of conflicts if it tries to merge at the same time as another rustdoc PR.

**Comment by tmandry** [comment]
@jyn514 rollups already ignore PRs that have merge conflicts when creating them

**Pull request merged**
